packages/mobx-react/src/observerClass.ts

```diff
@@ -50,142 +90,156 @@ export function makeClassComponentObserver(
+        // therefore `this` could NOT be garbage collected before mount,
+        // preventing reaction disposal by FinalizationRegistry and leading to memory leak.
+        // As an alternative we could have `admin.instanceRef = new WeakRef(this)`, but lets avoid it if possible.
+        admin.forceUpdate = () => this.forceUpdate()
+
+        if (!admin.reaction) {
+            // 1. Instance was unmounted (reaction disposed) and immediately remounted without running render #3395.
+            // 2. Reaction was disposed by finalization registry before mount. Shouldn't ever happen for class components:
+            // `componentDidMount` runs synchronously after render, but our registry are deferred (can't run in between).
+            // In any case we lost subscriptions to observables, so we have to create new reaction and re-render to resubscribe.
+            // The reaction will be created lazily by following render.
+            admin.forceUpdate()
         }
     })
-    patch(target, "componentWillUnmount", function () {
+
+    patch(prototype, "componentWillUnmount", function () {
         if (isUsingStaticRendering()) {
             return
         }
-
-        const reaction = this.render[mobxAdminProperty]
-        if (reaction) {
-            reaction.dispose()
-            // Forces reaction to be re-created on next render
-            this.render[mobxAdminProperty] = null
-        } else {
-            // Render may have been hot-swapped and/or overridden by a subclass.
-            const displayName = getDisplayName(this)
-            console.warn(
-                `The reactive render of an observer class component (${displayName})
-                was overridden after MobX attached. This may result in a memory leak if the
-                overridden reactive render was not properly disposed.`
-            )
-        }
-
-        this[mobxIsUnmounted] = true
+        const admin = getAdministration(this)
+        admin.reaction?.dispose()
+        admin.reaction = null
+        admin.forceUpdate = null
+        admin.mounted = false
     })
+
     return componentClass
 }
 
 // Generates a friendly name for debugging
-function getDisplayName(comp: any) {
-    return (
-        comp.displayName ||
-        comp.name ||
-        (comp.constructor && (comp.constructor.displayName || comp.constructor.name)) ||
-        "<component>"
-    )
+function getDisplayName(componentClass: ComponentClass) {
+    return componentClass.displayName || componentClass.name || "<component>"
 }
 
 function createReactiveRender(originalRender: any) {
-    /**
-     * If props are shallowly modified, react will render anyway,
-     * so atom.reportChanged() should not result in yet another re-render
-     */
-    setHiddenProp(this, skipRenderKey, false)
-    /**
-     * forceUpdate will re-assign this.props. We don't want that to cause a loop,
-     * so detect these changes
-     */
-    setHiddenProp(this, isForcingUpdateKey, false)
```

Comment:
Merged these together into `isUpdating` as they basically handle the same thing:
One is set by reaction to prevent reportChanged on `props` setter,
second is set by `props` setter to prevent reaction.
So both says "I am already updating, do not cause another update"